org.scala-ide.sdt.core/src/org/scalaide/core/internal/project/SbtScopesBuildManager.scala

```diff
@@ -0,0 +1,147 @@
+    new EclipseSbtBuildManager(owningProject, managerSettings, a, b)
+}
+
+private class ScopesBuildManagerFactory extends BuildManagerFactory {
+  def buildManager(owningProject: IScalaProject, managerSettings: Settings, a: Seq[IPath], b: Seq[(IContainer, IContainer)]): SbtAwareBuildManager =
+    new SbtScopesBuildManager(owningProject, managerSettings)
+}
+
+trait ScopeBuildManager extends EclipseBuildManager {
+  def hasErrors(scope: CompileScope): Boolean
+}
+
+class SbtScopesBuildManager(val owningProject: IScalaProject, managerSettings: Settings)
+    extends SbtAwareBuildManager {
+  @volatile private var error = false
+
+  val DefaultScopesOrder: Seq[CompileScope] = Seq(CompileMacrosScope, CompileMainScope, CompileTestsScope)
+
+  private val buildScopeUnits = DefaultScopesOrder.foldLeft(List.empty[BuildScopeUnit]) { (acc, scope) =>
+    scope match {
+      case CompileMacrosScope => new BuildMacrosUnit(owningProject, managerSettings) :: acc
+      case CompileMainScope => new BuildMainUnit(owningProject, managerSettings, acc.toSeq) :: acc
+      case CompileTestsScope => new BuildTestsUnit(owningProject, managerSettings, acc.toSeq) :: acc
+      case _ => acc
+    }
+  }.reverse
+
+  def hasErrors(compileScope: CompileScope): Boolean =
+    buildScopeUnits filter { _.scope == compileScope } exists { _.hasError }
+
+  override def hasErrors = error && (buildScopeUnits exists { _.hasError })
+
+  def build(addedOrUpdated: Set[IFile], removed: Set[IFile], monitor: SubMonitor): Unit = {
+    error = false
+    buildScopeUnits foreach { buildScopeUnit =>
+      if (areTransitiveDependenciesBuilt(buildScopeUnit)) {
+        buildScopeUnit.build(addedOrUpdated, removed, monitor)
+      } else {
+        error = true
+      }
+    }
+  }
+
+  def invalidateAfterLoad = true
+  def clean(implicit monitor: IProgressMonitor): Unit = buildScopeUnits.foreach { _.clean }
+  def canTrackDependencies: Boolean = true
+
+  private def areTransitiveDependenciesBuilt(buildScopeUnit: BuildScopeUnit) = {
+    def hasErrors(project: IProject, scope: CompileScope): Boolean = {
+      IScalaPlugin().getScalaProject(project).buildManager match {
+        case manager: SbtScopesBuildManager => manager.hasErrors(scope)
+        case manager: EclipseBuildManager => manager.hasErrors
+      }
+    }
+    val projectsInError = for {
+      scope <- buildScopeUnit.dependentScopesInUpstreamProjects
+      project <- owningProject.transitiveDependencies if hasErrors(project, scope)
+    } yield project
+
+    val stopBuildOnErrorsProperty = SettingConverterUtil.convertNameToProperty(ScalaPluginSettings.stopBuildOnErrors.name)
+    val stopBuildOnErrors = owningProject.storage.getBoolean(stopBuildOnErrorsProperty)
+
+    if (stopBuildOnErrors && projectsInError.nonEmpty) {
+      owningProject.underlying.deleteMarkers(SdtConstants.ProblemMarkerId, true, IResource.DEPTH_INFINITE)
+      BuildProblemMarker.create(owningProject.underlying,
```

Comment:
Looks like this code is duplicated, could we find a way to reduce this duplication?